# Spam Detector

In [17]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [18]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

Response: This question is not logical. We cannot figure this out at this point. That is the purpose of buidling the model. We do not understand how it will react to the mathmatical questions until we train and test most models, i.e., run the program. That said: 

Logistic regression would not be the best model here. The data given is more complex. 

Random Forest Classifier (RFC) would be best suited and the best performer here because RFC can:
- Can model complex relationships between variables.
- Can handle large amounts of data efficiently.

Random Forests are known for handling complex relationships in data and are generally more robust to overfitting compared to Logistic Regression.

The bank marketing dataset might have non-linear relationships between features and the target variable, which Random Forests can capture effectively.

Random Forests can handle a larger number of features well without the need for feature scaling, which might be beneficial in this dataset.

## Split the Data into Training and Testing Sets

In [19]:
# Create the labels set `y` and features DataFrame `X`
# Create the labels set 'y' from the 'spam' column
y = data['spam']

# Create the features DataFrame 'X' from the remaining columns
X = data.drop(columns=['spam'])

In [20]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
y_balance = y.value_counts()
y_balance

spam
0    2788
1    1813
Name: count, dtype: int64

In [21]:
# Split the data into X_train, X_test, y_train, y_test

X_train, X_test, y_train, y_test = train_test_split(X, y,test_size=0.2, random_state=42)

X_train.head()



word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
1370            0.09                0.0           0.09           0.0   
3038            0.00                0.0           0.00           0.0   
2361            0.00                0.0           2.43           0.0   
156             0.00                0.0           0.00           0.0   
2526            0.00                0.0           0.00           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
1370           0.39            0.09              0.09                0.00   
3038           0.00            0.00              0.00                0.00   
2361           0.00            0.00              0.00                0.00   
156            1.31            0.00              1.31                1.31   
2526           0.00            0.00              0.00                0.00   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
1370             0.19            0.29  ...                   0.0        0.000   
3038             0.00            0.00  ...                   0.0        0.124   
2361             0.27            0.00  ...                   0.0        0.000   
156              1.31            1.31  ...                   0.0        0.000   
2526             0.00            0.00  ...                   0.0        0.000   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
1370        0.139          0.0        0.310        0.155          0.0   
3038        0.124          0.0        0.000        0.000          0.0   
2361        0.344          0.0        0.000        0.000          0.0   
156         0.000          0.0        0.117        0.117          0.0   
2526        0.000          0.0        0.000        0.000          0.0   

      capital_run_length_average  capital_run_length_longest  \
1370                       6.813                         494   
3038                       1.800                           8   
2361                       2.319                          12   
156                       48.500                         186   
2526                       2.307                           8   

      capital_run_length_total  
1370                      1458  
3038                        45  
2361                       167  
156                        291  
2526                        30  

[5 rows x 57 columns]

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [22]:


# Create the StandardScaler instance
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()   


In [23]:
# Fit the Standard Scaler with the training data
scaler = StandardScaler().fit(X_train)


In [24]:
# Scale the training data
X_train_scaled= scaler.transform(X_train)

#we should also add scale the testing data to show if model if overfit or not

X_test_scaled =scaler.transform(X_test)


## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [25]:
# Train a Logistic Regression model and print the model score
# move to import list from sklearn.linear_model import LogisticRegression


Log_model =LogisticRegression(random_state =1)
Log_model.fit(X_train_scaled, y_train)
Log_model.score

accuracy = Log_model.score(X_test_scaled, y_test)

print("Accuracy of the logistic regression model: {:.3f}".format(accuracy))



Accuracy of the logistic regression model: 0.920


In [26]:
# Make and save testing predictions with the saved logistic regression model using the test data
    ##given
    ### this is where we name our model
    #given array([1, 0, 1, ..., 0, 0, 0])
# Review the predictions

import joblib
joblib.dump(Log_model,"saved_logistic_regression_model.pkl") ## this is me saving the model 

# Load the saved logistic regression model
saved_model = joblib.load('saved_logistic_regression_model.pkl')

# Make predictions on the test data
testing_predictions = saved_model.predict(X_test_scaled)

# Save the predictions
joblib.dump(testing_predictions, 'testing_predictions.pkl')
# In this code snippet:
# joblib.dump(testing_predictions, 'testing_predictions.pkl') #saves the predictions in a file named testing_predictions.pkl.
# joblib.load('saved_logistic_regression_model.pkl') loads the saved logistic regression model.
# saved_model.predict(X_test_scaled) makes predictions on the test data.

print(testing_predictions)

[0 0 0 1 0 1 0 0 0 1 0 0 0 1 0 0 0 0 1 0 0 1 0 0 1 0 1 0 1 1 1 1 0 1 0 0 1
 1 0 1 1 0 0 0 0 1 1 0 0 1 1 0 0 0 0 1 0 0 0 0 1 0 1 1 1 0 0 0 0 0 0 1 0 1
 1 0 0 0 0 1 0 0 0 0 0 0 1 1 1 0 1 1 0 0 0 0 1 0 1 1 0 1 0 1 1 0 0 1 1 1 1
 1 1 0 1 1 0 0 0 0 1 0 1 0 1 0 0 0 1 0 0 1 1 0 0 0 1 0 1 1 1 1 0 0 1 1 0 0
 0 0 0 1 1 1 0 0 0 1 1 0 1 1 1 0 1 0 1 0 0 1 0 0 0 0 1 0 1 0 0 1 0 0 0 0 1
 1 0 0 0 0 1 0 1 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 1 0 1 0 0 0 0 0 0 0 0 0 1
 1 1 1 0 1 0 0 0 0 0 0 0 0 0 0 0 1 1 0 1 1 1 0 0 0 1 0 0 0 1 0 0 0 0 0 1 1
 0 1 0 1 0 1 1 1 0 0 0 1 0 0 0 0 0 1 0 0 0 0 1 0 0 1 0 0 1 0 1 0 1 0 0 1 1
 0 0 0 0 0 1 0 0 0 0 1 1 1 0 1 0 1 1 0 0 0 1 1 0 0 0 0 0 0 1 1 1 1 0 0 0 1
 0 0 1 0 1 0 0 0 0 0 0 1 0 0 0 1 1 0 1 0 0 0 1 0 1 1 1 0 1 1 0 0 0 0 0 0 0
 1 1 0 1 0 0 0 0 1 0 1 1 0 0 1 0 1 0 0 0 0 0 1 1 0 1 0 1 0 0 1 0 0 0 1 1 0
 1 1 1 0 0 1 0 0 1 0 0 0 1 0 1 0 0 0 0 1 1 0 0 0 1 0 0 1 1 0 0 1 0 0 0 1 1
 0 1 0 0 0 1 0 0 0 1 0 0 1 1 1 1 1 1 0 0 1 0 1 1 1 1 0 0 0 0 0 0 0 0 0 0 1
 0 1 1 1 0 0 0 0 0 1 1 1 

In [27]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.

testing_predictions = joblib.load('testing_predictions.pkl')
accuracy = accuracy_score(y_test, testing_predictions)
print(f"Accuracy Score: {accuracy}")


Accuracy Score: 0.9196525515743756


## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [28]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier


In [29]:

# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier
# Make and save testing predictions with the saved logistic regression model using the test data

rf_model = RandomForestClassifier(random_state=43)
rf_model.fit(X_train_scaled, y_train)

y_pred = rf_model.predict(X_test_scaled)

# Evaluate the model's performance by calculating the accuracy score
accuracy = accuracy_score(y_test, y_pred)

# Review the predictions
print(f"Model Accuracy: {accuracy}")

Model Accuracy: 0.9565689467969598


In [30]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy = accuracy_score(y_test, testing_predictions)
print(f"Accuracy Score: {accuracy}")

Accuracy Score: 0.9196525515743756


## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

RESPONSE: 
The Random Forest Classifier performed better with an accuracy score of 95.66%, compared to the Logistic Regression model which had an accuracy of 92.0%. Our prediction was that the Random Forest Classifier would preform better due to the data set we were working with, as it happens to have a complex data set. 